### ISSUE TITLE
Copying/pasting node creates duplicate children

### ISSUE THREAD

**Issue by Pennycook** [godotengine/godot#62614]
### Godot version

3.4.4.stable.official

### System information

Ubuntu 20.04

### Issue description

I'm seeing strange behavior when copying/pasting a custom node that creates its children inside of `_init`, like so:
```gdscript
func _init():
	var child = preload("res://addons/dummy_plugin/CustomScene.tscn").instance()
	child.owner = null
	add_child(child)
```

When the node is added to the scene initially, it appears in the scene dock like this, without the children shown:
![custom-type](https://user-images.githubusercontent.com/857660/176968851-8463fd0f-0264-47ad-8184-0f5caa94841f.png)

If I copy and paste the node, the copy has twice as many children as expected, and they appear in the scene dock:
![custom-type-children](https://user-images.githubusercontent.com/857660/176969058-2128a046-b994-4c0c-b0ad-83a278d355f0.png)

### Steps to reproduce

1. Open Test.tscn
2. Add a CustomType as a child of the root Control node
3. Copy and paste the node

### Minimal reproduction project

[custom-type-children.zip](https://github.com/godotengine/godot/files/9031615/custom-type-children.zip)

**Comment by snailrhymer** [comment]
Also occurs in `master`. Note that in the example above, the node `CustomType2` actually has three children, with one that correctly has `owner = null` and so doesn't appear in the scene tree dock.

I've found the things causing this behavior:

1. [Pasting automatically sets children's owners as their parents](https://github.com/godotengine/godot/blob/d94253271011e9fa4f74fdbd02991732c0d8f8e6/editor/scene_tree_dock.cpp#L3083), so any child nodes that weren't owned by their parents will be owned in the duplicated version, and so appear in the Scene Tree Dock.

2. `_init` runs each time a node is duplicated. Copying\* a node duplicates it to the "node clipboard", and pasting a node duplicates it a second time from the clipboard back to the scene tree.

    (\* Since [cutting is implemented as a special case of copying](https://github.com/godotengine/godot/blob/d94253271011e9fa4f74fdbd02991732c0d8f8e6/editor/scene_tree_dock.cpp#L447), `_init` will also be run twice on a node each time it is cut + pasted, which may be particularly counter-intuitive.)

Together, these mean that when the original `CustomType` node is copied, it is duplicated to the clipboard, `_init` is called and it gains a child with no owner. Then when it is pasted, `_init` is called on the new node in the scene tree and an unowned child is added. In a later step in the pasting, the two children are added from the clipboard version, but now they have `owner` set to their parent.

***

#63130 offers a fix for (1).

I'm not sure what desirable behavior for (2) would be.

**Comment by snailrhymer** [comment]
I think you'd expect similar results with `add_child` moved almost anywhere - the problem is that duplicating copies a node and all its children, and the `add_child` logic always adds a child, so the number of children will always increase with each duplicate. One way to fix this is to not always add a child:

```gdscript
tool
extends Control

const CHILD_NAME = "SomeName"

func _ready():
	if not has_node(CHILD_NAME):
		var child = preload("res://addons/dummy_plugin/CustomScene.tscn").instance()
		child.owner = null
		add_child(child)
		child.name = CHILD_NAME
	print(get_child_count())

```
though note that this can't happen earlier than `_ready()`, or else the child you're checking for may not have been added.

It might be that there's another nice way to do what you're doing - I'm relatively new to working with the engine.

**Comment by Pennycook** [comment]
> It might be that there's another nice way to do what you're doing - I'm relatively new to working with the engine.

I'm new to this sort of thing, too -- I stumbled across this bug while experimenting with custom types, which I've very little experience with.  I was led in this direction by some of the documentation, which shows `add_child` being called in functions like `_init` and `_ready`.

Also, the [documentation](https://docs.godotengine.org/en/stable/classes/class_node.html?highlight=Node#class-node-property-owner) for the `owner` property of `Node` says:
> Note: If you want a child to be persisted to a [PackedScene](https://docs.godotengine.org/en/stable/classes/class_packedscene.html#class-packedscene), you must set [owner](https://docs.godotengine.org/en/stable/classes/class_node.html?highlight=Node#class-node-property-owner) in addition to calling [add_child](https://docs.godotengine.org/en/stable/classes/class_node.html?highlight=Node#class-node-method-add-child). This is typically relevant for [tool scripts](https://docs.godotengine.org/en/3.4/tutorials/misc/running_code_in_the_editor.html) and [editor plugins](https://docs.godotengine.org/en/3.4/tutorials/plugins/editor/index.html). If [add_child](https://docs.godotengine.org/en/stable/classes/class_node.html?highlight=Node#class-node-method-add-child) is called without setting [owner](https://docs.godotengine.org/en/stable/classes/class_node.html?highlight=Node#class-node-property-owner), the newly added Node will not be visible in the scene tree, though it will be visible in the 2D/3D view.

...which suggests that it's intentional that there is a way to create children without them showing up in the editor, and that the duplication that happens here is not expected.

I agree that your `CHILD_NAME` idea works, but it relies on nobody else creating a node called `CHILD_NAME`. That might not be a safe assumption when the custom type is shipped as an editor plugin.

Wandering into proposal territory here, but perhaps there should be some mechanism (like a notification) that tells a Node it's about to be cut/copied? If it was possible to detect that somehow, the script could clean up anything it didn't want to be copied, let the copy happen, and then re-run the `_init`/`_enter_tree`/`_ready` sequence as normal.

**Comment by Zireael07** [comment]
> ...which suggests that it's intentional that there is a way to create children without them showing up in the editor, and that the duplication that happens here is not expected.

Yes and yes

**Comment by snailrhymer** [comment]
> I agree that your `CHILD_NAME` idea works, but it relies on nobody else creating a node called `CHILD_NAME`. That might not be a safe assumption when the custom type is shipped as an editor plugin.

Good point. I suppose you could go for a name like `_this_name_is_only_used_by_MYPLUGIN_dont_name_any_of_your_nodes_this`, but yeah not ideal.

> What if nodes with `owner` set to `null` were just never copied into the clipboard at all?

This might cause an inverse problem - if I have a tool script that adds unowned child nodes to a bunch of _other_ nodes (e.g. labels above units or something), it could be frustrating to not have the children copied when I copy one of the parents.

There's also that [duplication](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-method-duplicate) is something nodes can do in the engine itself, and it could be very confusing to have differing functionality between `Node.duplicate()` and the `Copy/Paste` or `Duplicate` editor tools. 

> perhaps there should be some mechanism (like a notification) that tells a Node it's about to be cut/copied?

This could work, but might need users to be familiar with how copy/pasting and duplication works (e.g. the `Duplicate` tool only instantiates a node and its children once, while `Copy > Paste` does it twice). Something simpler like a `do_not_duplicate` flag on nodes might be preferable. Either way, it's probably worth a proposal to get more thoughts on it.

**Comment by Pennycook** [comment]
> Good point. I suppose you could go for a name like `_this_name_is_only_used_by_MYPLUGIN_dont_name_any_of_your_nodes_this`, but yeah not ideal.

What a great variable name. :laughing:

> This might cause an inverse problem - if I have a tool script that adds unowned child nodes to a bunch of _other_ nodes (e.g. labels above units or something), it could be frustrating to not have the children copied when I copy one of the parents.

Good point, I hadn't considered that.

> There's also that [duplication](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-method-duplicate) is something nodes can do in the engine itself, and it could be very confusing to have differing functionality between `Node.duplicate()` and the `Copy/Paste` or `Duplicate` editor tools.

I didn't realize that `duplicate()` worked this way.  I guess that means that this issue isn't Editor-specific... Any node that creates children inside of a function like `_init` will create extra copies of those children when `duplicate()` is called in a script, too.

> This could work, but might need users to be familiar with how copy/pasting and duplication works (e.g. the `Duplicate` tool only instantiates a node and its children once, while `Copy > Paste` does it twice). Something simpler like a `do_not_duplicate` flag on nodes might be preferable. Either way, it's probably worth a proposal to get more thoughts on it.

That's true. An ideal solution would ensure that the node behaved as the user expected regardless of how many times it ended up being duplicated, and something like a `do_not_duplicate` flag would handle that.

@Calinou, is there anybody you'd recommend talking to about the design of `duplicate()` before we open a proposal?  Because there are instances of `add_child` appearing inside `_init` in the documentation (e.g. [here](https://docs.godotengine.org/en/stable/tutorials/best_practices/scenes_versus_scripts.html?highlight=add_child#performance-of-script-vs-packedscene)), it feels like somebody might have thought about the safety of this pattern before. I want to make sure we're not overlooking a solution that already exists.

**Comment by Calinou** [comment]
> @Calinou, is there anybody you'd recommend talking to about the design of duplicate() before we open a proposal?

Probably @reduz.

**Comment by Pennycook** [comment]
> Probably @reduz.

Great, thanks.

After some digging I found https://github.com/godotengine/godot/issues/3393, where @reduz preempted this issue and [said](https://github.com/godotengine/godot/issues/3393#issuecomment-307800787):
> duplicate() for nodes is hackish, and not generally recommended ... The reason is that some nodes create subnodes (ie, a list creates it's scrollbars), so a simple duplicate() function has to be able to tell to a certain degree what to duplicate and what not to duplicate. This is not always obvious.

...The `do_not_duplicate` property suggestion seems aligned with the thinking here.

However, that issue then led me to https://github.com/godotengine/godot/pull/57121, a draft PR from @KoBeWi that closes #3393 by making `duplicate()` copy a node's internal state by default.  If I'm understanding it correctly, this could also provide a solution to the problem by allowing scripts to keep track of whether they've already created their children:

```gdscript
tool
extends Control

var _initialized := false

func _enter_tree():
	if not _initialized:
		var child = preload("res://addons/dummy_plugin/CustomScene.tscn").instance()
		child.owner = null
		add_child(child)
		_initialized = true
```

@reduz, do you have any thoughts/preference here?  Personally, I think leveraging #57121 would be simpler than adding a new property to each child, but #57121 has been open for a while and I'm not sure if it will be merged.

**Comment by KoBeWi** [comment]
I can't reproduce the bug in 4.0 beta 2 using the attached MRP.

**Comment by KoBeWi** [comment]
Fixed in 4.1

**Comment by kleonc** [comment]
> Fixed in 4.1

@KoBeWi Not really, seems like only (1) from https://github.com/godotengine/godot/issues/62614#issuecomment-1186620513 is fixed (by #63130). The nodes are still being duplicated though, they're just not visible in the editor's scene tree dock as no owner is set for them.

Try e.g. such scene (inspired by #77404):
![godot windows editor dev x86_64_OQwx8myLl3](https://github.com/godotengine/godot/assets/9283098/e03e4bbe-5c84-41e8-9400-2c9f36ca6bb0)
with script:
```gdscript
@tool
extends PathFollow2D

func _ready() -> void:
	prints(self, get_children(true))
```
After few duplicates:
![godot windows editor dev x86_64_UDdEZcNXsm](https://github.com/godotengine/godot/assets/9283098/c24dfc56-4133-4f9b-95a1-8facc0d14af2)
```
PathFollow2D2:<PathFollow2D#3491422604232> [_Timer_21329:<Timer#3491439380933>]
Duplicate Node(s)
PathFollow2D3:<PathFollow2D#3532375788942> [_Timer_21329:<Timer#3532392565634>, _Timer_38237:<Timer#3532409343391>]
Duplicate Node(s)
PathFollow2D4:<PathFollow2D#3551954799638> [_Timer_21329:<Timer#3551971576314>, _Timer_38237:<Timer#3551988354972>, _Timer_38946:<Timer#3552005132255>]
Duplicate Node(s)
```
Copy-pasting is also affected (but you'd need to copy anew after each pasting to replicate the issue with duplicating).

**Comment by KoBeWi** [comment]
Maybe open a new issue, because it's lesser severity. The duplicate nodes don't appear in scene tree and are temporary.
